---
title: "Empirical data analysis"
author: "Somya Mehra"
date: "2024-02-04"
output: html_document
---

```{r setup, include=FALSE}
library(paneljudge)
library(poisbinom)
library(Pv3Rs)
library(paneljudge)
library(igraph)
library(dplyr)
library(tidyr)
library(readr)
library(ggplot2)
library(cowplot)
library(ggridges)
library(ggh4x)
library(patchwork)
library(viridis)

knitr::opts_chunk$set(echo=TRUE, warning=FALSE, message=FALSE, fig.align = "center")

library(extrafont)
font_import() # takes a few minutes
loadfonts(device="postscript")

library(showtext)
font_add(family = "Arial", regular = "Arial.ttf") ## here is the path to the font to add.
showtext.auto()
```

## Data input

```{r data_input}
set.seed("260823")

source("Simulation_model_code/estimate_relatedness.R")

if (!dir.exists("Empirical_results")) dir.create("Empirical_results")

RUN_ALL <- FALSE

# WGS data for Guyana and Colombia (combined)
empirical_WGS_data <- read_rds("Empirical_data/WGS_genotype_matrix.rds")
isolate_metadata <- read_rds("Empirical_data/WGS_isolate_metadata.rds")
n_isolates <- nrow(empirical_WGS_data[["genotype_matrix"]])
n_variants <- ncol(empirical_WGS_data[["genotype_matrix"]])

RHO <- 7.4 * 10^(-7)
N_MARKER_DOWNSAMPLE <- c(100, 500, 2000, 8000)
IBD_THRESHOLDS <- seq(0.25, 0.975, 0.025)

# Simulated data (for generating composite elbow plot figure)
dense_data_sim <- read_rds("Numerical_results/Simulated_data/dense_data_simulation.rds")
dense_data_sim_ibd <- read_rds("Numerical_results/Simulated_data/dense_data_sim_ibd_estimates.rds")
GENERATION <- 10
```

## Guyana WGS data

We first analyse the set of `r sum(isolate_metadata$Country=="Guyana")` (deemed to be monoclonal) isolates sampled from Guyana. 

```{r Guyana_genotypes}
Guyana_isolates <- subset(isolate_metadata$ID, isolate_metadata$Country=="Guyana")
Guyana_genotypes <- empirical_WGS_data[["genotype_matrix"]][Guyana_isolates,]
polymorphic_SNPs <- which(colSums(Guyana_genotypes==0, na.rm=TRUE)>1 & 
                            colSums(Guyana_genotypes==1, na.rm=TRUE)>1)
Guyana_genotypes <- Guyana_genotypes[, polymorphic_SNPs]
Guyana_variants <- empirical_WGS_data[["positions"]][polymorphic_SNPs,]
n_Guyana_variants <- length(polymorphic_SNPs)
```

We remove variants that are monomorphic in the Guyana to yield `r n_Guyana_variants` biallelic SNP sites. We then estimate pairwise relatedness based on these dense data for this sample.

```{r Guyana_ibd}
if (RUN_ALL || !file.exists("Empirical_results/Guyana_pairwise_relatedness.rds")) {
  Guyana_ibd <- generate_relatedness_summary(Guyana_genotypes, Guyana_variants, rho=RHO)
  write_rds(Guyana_ibd, "Empirical_results/Guyana_pairwise_relatedness.rds", compress="gz")
} else {
  Guyana_ibd <- read_rds("Empirical_results/Guyana_pairwise_relatedness.rds")
}
```

We also estimate pairwise relatedness for sparse data, generated by downsampling subsets of SNPs uniformly at random (without replacement).

```{r Guyana_ibd_thinned}
downsample_marker_subsets <- lapply(N_MARKER_DOWNSAMPLE, function(n_subset) {
  sort(sample(n_Guyana_variants, n_subset, replace=FALSE))})

names(downsample_marker_subsets) <- paste0("N_", N_MARKER_DOWNSAMPLE)

downsampled_data_ibd <- list()
for (i in names(downsample_marker_subsets)) {
  if (RUN_ALL || !file.exists(paste0("Empirical_results/Guyana_downsampled_", i, ".rds"))) {
    markers <- downsample_marker_subsets[[i]]
    downsampled_data_ibd[[i]] <- generate_relatedness_summary(Guyana_genotypes[, markers],
                                                              Guyana_variants[markers,], rho=RHO)
    write_rds(downsampled_data_ibd[[i]], paste0("Empirical_results/Guyana_downsampled_", i, ".rds"), compress="gz")
  }
  else {
    downsampled_data_ibd[[i]] <- read_rds(paste0("Empirical_results/Guyana_downsampled_", i, ".rds"))
  }
}
```

Relatedness estimates predicated on dense WGS data are treated as the gold-standard.

```{r dense_data_gold_standard}
WGS_thinned_vs_all <- lapply(downsampled_data_ibd, function(x) {
  x[["summary_metrics"]] %>%
    merge(Guyana_ibd[["summary_metrics"]], by=c("V1", "V2"), 
          suffixes = c(".thinned",".all"))}) %>%
  bind_rows(.id="n_markers") %>%
  mutate(n_markers=factor(gsub("N_", "No. SNPs: ", n_markers),
                          levels=paste0("No. SNPs: ", N_MARKER_DOWNSAMPLE)))

threshold_classification <- lapply(IBD_THRESHOLDS, function(threshold) {
  classification_summary <- WGS_thinned_vs_all %>% 
    mutate(thinned_class=rhat_hmm_allele.thinned>=threshold, 
           WGS_class=rhat_hmm_allele.all>=threshold) %>% 
    group_by(thinned_class, WGS_class, n_markers) %>% 
    summarise(count=n()) %>%
    mutate(type=ifelse(thinned_class & WGS_class, "TP",
                       ifelse(thinned_class, "FP", 
                              ifelse(WGS_class, "FN", "TN")))) %>%
    as.data.frame %>% select(-thinned_class, -WGS_class) %>% 
    reshape2::dcast(n_markers~type, value.var="count") 
  classification_summary[is.na(classification_summary)] <- 0
  
  cols_to_create <- setdiff(c("TN", "FN", "TP", "FP"), 
                            colnames(classification_summary))
  classification_summary[cols_to_create] <- 0
  
  classification_summary <- classification_summary %>%
    transmute(n_markers=n_markers, threshold=threshold,
              sensitivity=TP/(TP+FN), specificity=TN/(FP+TN),
              PPV=TP/(TP+FP), NPV=TN/(FN+TN))
  
  return(classification_summary)}) %>%
  do.call(rbind, .) %>% as.data.frame

```

```{r Guyana_summary_plot, echo=FALSE, fig.width=9, fig.height=7.2}
sparse_indep_plots <-   ggplot(WGS_thinned_vs_all) +
  geom_point(aes(x=rhat_hmm_allele.all, y=rhat_indep_allele.thinned), size=0.01, alpha=0.1) +
  geom_abline(lty=2, col="navyblue") +
  facet_grid(cols=vars(n_markers)) + 
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) + 
  xlab("Dense WGS data estimate under HMM") +
  ylab("Thinned data estimates\nunder independence") +
  theme_bw()  

sparse_hmm_plots <-   ggplot(WGS_thinned_vs_all) +
  geom_point(aes(x=rhat_hmm_allele.all, y=rhat_hmm_allele.thinned), size=0.01, alpha=0.1) +
  geom_abline(lty=2, col="navyblue") +
  facet_grid(cols=vars(n_markers)) + 
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) + 
  xlab("Dense WGS data estimate under HMM") +
  ylab("Thinned data estimates\nunder HMM") +
  theme_bw()  

classification_plot <- 
  ggplot(threshold_classification, aes(x=threshold, y=sensitivity)) + 
  geom_line(col="darkgray") + geom_point(size=0.8) + facet_grid(cols=vars(n_markers)) + 
  xlab("Relatedness threshold") + ylab("Classification sensitivity\n(under HMM)") +
  coord_cartesian(ylim=c(0, 1)) + theme_bw()

Guyana_summary_plot <-
  cowplot::plot_grid(sparse_indep_plots, sparse_hmm_plots, classification_plot, 
                   ncol=1, align="vh", axis="tblr", labels=c("(A)", "(B)", "(C)"))

show(Guyana_summary_plot)

png("Empirical_results/Guyana_summary.png", width=9, height=7.2, units="in", res=300)
show(Guyana_summary_plot)
invisible(dev.off())
```

## Elbow plots

We collate the dense data diagnostic for Guyana with our theoretical and numeric results (which do not account for population structure)

```{r elbow_plot, echo=FALSE, fig.width=13.5, fig.height=4.5}
elbow_plots <- list()
locuswise_ibd_plots <- list()

elbow_plots[["theoretical"]] <- ggplot() +
  geom_point(aes(x=Inf, y=Inf)) +
  #geom_polygon(aes(x=c(0, 1, 0.5), y=c(0, 1, 0), group=c(1, 1, 1)), alpha=0.2) +
  geom_segment(aes(x=c(0, 0.25), xend=c(0.25, 1), y=c(0, 0), yend=c(0, 1)), 
               col="black", lwd=1.4) +
  geom_vline(aes(xintercept=0.25), color="#9e8fc4", lwd=0.8) +
  geom_segment(aes(x=0, xend=1, y=0, yend=1), color="darkgray", lty=2, lwd=0.6) + 
  annotate("label", x=0.25, y=0.9, 
           label="Mean locuswise\nproportion of\nIBD pairs", fill="#9e8fc4", col="white") +
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) + 
  xlab("Relatedness estimates under independence\nwith theoretically corrected nIBD-to-IBS model") +
  ylab("Relatedness estimates under independence\nwith standard nIBD-to-IBS model") +
  theme_light()

elbow_plots[["theoretical"]] <- 
  plot_spacer() + plot_spacer() + elbow_plots[["theoretical"]] + plot_spacer() +
  plot_layout(ncol = 2, nrow = 2, widths = c(4, 0.2), heights = c(1, 4)) +
  plot_annotation(title = '(A) Theoretical',
                  theme = theme(plot.title = element_text(size = 12, hjust=0.5, face="bold"),
                                plot.background = element_rect(fill="grey"),
                                plot.margin = unit(c(0.2,0,0,0), "cm")))


# inbreeding_multigen[["locus_metrics"]] %>% bind_rows(.id="generation") %>% mutate(maf=1/2-sqrt(2*prop_pair_ibs-1)/2) %>% ggplot() + geom_histogram(aes(x=maf), binwidth=0.01) + facet_wrap(vars(generation))

# ============================ NUMERICAL ESTIMATES =====================

sim_ibd_estimates <- dense_data_sim_ibd[["ibd_ibs_est"]] %>% 
  subset(generation==GENERATION)

sim_mean_locuswise_IBD <- 
  mean((dense_data_sim[["locus_metrics"]][[GENERATION]] %>% subset(prop_pair_ibs<1))$prop_pair_ibd)

elbow_plots[["numerical"]] <- ggplot(sim_ibd_estimates) + 
  geom_point(aes(x=realised_IBD, y=rhat_naive_indep), size=0.2, alpha=0.3) + 
  geom_abline(color="darkgray", lty=2) +
  #scale_x_continuous(expand=c(0, 0)) +
  #scale_y_continuous(expand=c(0, 0)) +
  geom_vline(aes(xintercept=sim_mean_locuswise_IBD), color="#9e8fc4", lwd=0.8) +
  annotate("label", x=sim_mean_locuswise_IBD, y=0.9, 
           label="Mean locuswise\nproportion of\nIBD pairs", fill="#9e8fc4", col="white") +
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) + 
  xlab("Realised fraction of sites IBD\n") +
  ylab("Relatedness estimates under independence\nwith standard (n)IBD-to-allele model") +
  theme_light()

numerical_x_hist <-  ggplot(sim_ibd_estimates) + 
  geom_histogram(aes(x=realised_IBD), bins=100, col="black", lwd=0.25) + 
  theme_void() + theme(plot.background = element_rect(fill="white", size=0))

numerical_y_hist <-  ggplot(sim_ibd_estimates) + 
  geom_histogram(aes(x=rhat_naive_indep), bins=100, col="black", lwd=0.25) + 
  coord_flip() + 
  theme_void() + theme(plot.background = element_rect(fill="white", size=0))

elbow_plots[["numerical"]] <- 
  numerical_x_hist + plot_spacer() + elbow_plots[["numerical"]] + numerical_y_hist +
  plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4)) +
  plot_annotation(title = '(B) Numerical',
                  theme = theme(plot.title = element_text(size = 12, hjust=0.5, face="bold"),
                                plot.background = element_rect(fill="grey"),
                                plot.margin = unit(c(0.2,0,0,0), "cm")))


# ============================ EMPIRICAL ESTIMATES =====================
elbow_plots[["empirical"]] <- ggplot(Guyana_ibd[["summary_metrics"]]) + 
  geom_point(aes(x=rhat_hmm_allele, y=rhat_indep_allele), size=0.01, alpha=0.1) + 
  geom_abline(color="darkgray", lty=2) +
  #scale_x_continuous(expand=c(0, 0)) +  
  #scale_y_continuous(expand=c(0, 0)) +
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) + 
  xlab("Relatedness estimates under the HMM\nwith standard (n)IBD-to-allele model") +
  ylab("Relatedness estimates under independence\nwith standard (n)IBD-to-allele model") +
  theme_light()

empirical_x_hist <-  ggplot(Guyana_ibd[["summary_metrics"]]) + 
  geom_histogram(aes(x=rhat_hmm_allele), bins=100, col="black", lwd=0.25) + 
  theme_void() + theme(plot.background = element_rect(fill="white", size=0))

empirical_y_hist <-  ggplot(Guyana_ibd[["summary_metrics"]]) + 
  geom_histogram(aes(x=rhat_indep_allele), bins=100, col="black", lwd=0.25) + 
  coord_flip() + 
  theme_void() + theme(plot.background = element_rect(fill="white", size=0))

elbow_plots[["empirical"]] <- 
  empirical_x_hist + plot_spacer() + elbow_plots[["empirical"]] + empirical_y_hist +
  plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4)) +
  plot_annotation(title = '(C) Empirical',
                  theme = theme(plot.title = element_text(size = 12, hjust=0.5, face="bold"),
                                plot.background = element_rect(fill="grey"),
                                plot.margin = unit(c(0.2,0,0,0), "cm")))


# ============================ PANEL PLOT =====================

elbow_plots <- cowplot::plot_grid(elbow_plots[["theoretical"]], NULL, 
                                  elbow_plots[["numerical"]], NULL, elbow_plots[["empirical"]], 
                                  align="hv", axis="tblr", nrow=1, 
                                  rel_widths = c(4.2, 0.2, 5, 0.2, 5))
show(elbow_plots)

png("Numerical_results/Figures/elbow_plots.png",width=13.5, height=4.5, units="in", res=500)
show(elbow_plots)
invisible(dev.off())
```


## Diagnostics for population structure

We now consider the complete sample of `r n_isolates` parasites (deemed to be monoclonal) from both Guyana and Colombia, genotyped at `r n_variants` biallelic SNP sites. We begin by generating a summary of pairwise relatedness and IBS.

```{r full_pop_ibd}
if (RUN_ALL || !file.exists("Empirical_results/Guyana_Colombia_pairwise_relatedness.rds")) {
  full_pop_ibd <- generate_relatedness_summary(empirical_WGS_data[["genotype_matrix"]],
                                               empirical_WGS_data[["positions"]], rho=RHO)
  write_rds(full_pop_ibd, "Empirical_results/Guyana_Colombia_pairwise_relatedness.rds",
            compress="gz")
} else {
  full_pop_ibd <- read_rds("Empirical_results/Guyana_Colombia_pairwise_relatedness.rds")
}

```

### Principal coordinates analysis

We perform principal coordinates analysis, using the proportion of nIBS sites as a distance metric and distinguish subpopulations on the basis of the first principal coordinate

```{r hamming_pcoa}
n_pairs <- n_isolates*(n_isolates-1)/2
pairwise_ibs <- full_pop_ibd[["summary_metrics"]][,c("V1", "V2", "ibs")]

isolate_IDs <- rownames(empirical_WGS_data[["genotype_matrix"]])

# perform PCoA based on hamming distance
hamming_distance <- matrix(0, nrow=n_isolates, ncol=n_isolates,
                           dimnames = list(isolate_IDs, isolate_IDs))

for (i in 1:n_pairs) {
  hamming_distance[pairwise_ibs[i, "V1"], pairwise_ibs[i, "V2"]] <- 
      hamming_distance[pairwise_ibs[i, "V2"], pairwise_ibs[i, "V1"]] <- 1 - pairwise_ibs[i, "ibs"]
}

hamming_pcoa <- cmdscale(hamming_distance, k=3) %>% as.data.frame %>%
  mutate(Country=isolate_metadata[rownames(.), "Country"])
```

```{r hamming_pcoa_plot, echo=FALSE, fig.height=4, fig.width=6}
pop_comparison <- c("grey", "navyblue", "#dec268", "#49995b")

PCoA_plot <- ggplot(hamming_pcoa) + 
  geom_point(aes(x=V1, y=V2, fill=Country), pch=21, color="black") + 
  scale_fill_manual(values=c("#dec268", "navyblue"), name="Country") +
  xlab("Principal coordinate 1") + 
  ylab("Principal coordinate 2") + 
  ggtitle("PCoA based on fraction of sites IBS") +  
  theme_bw() + theme(plot.title = element_text(hjust=0.5, face="bold"))

show(PCoA_plot)

pdf("Empirical_results/Guyana_Colombia_PCoA.pdf", height=4, width=6)
show(PCoA_plot)
invisible(dev.off())
```

### Proportion of pairs IBS

We extract the pairwise fraction of sites IBS as diagnostic for population structure.

```{r ibs_diagnostic_plot, echo=FALSE, fig.height=2.5, fig.width=10}
summary_metrics <- full_pop_ibd[["summary_metrics"]] %>%
  mutate(comparison=ifelse(isolate_metadata[V1, "Country"]==isolate_metadata[V2, "Country"], 
                           isolate_metadata[V1, "Country"], "Guyana-Colombia"))

summary_metrics <- bind_rows(summary_metrics,
                             summary_metrics %>% mutate(comparison="Overall")) %>%
  mutate(comparison=factor(comparison, levels=c("Overall", "Guyana", 
                                                "Colombia", "Guyana-Colombia")))

ibs_diagnostic_plot <- ggplot(summary_metrics) + 
  geom_histogram(aes(x=ibs, fill=comparison), bins=100, col="black", lwd=0.1) + 
  facet_grid2(cols=vars(comparison), scales="free_y", independent="y") +
  scale_fill_manual(values=pop_comparison, name="Comparison") +
  xlab("Fraction of polymorphic sites IBS") + ylab("Frequency") +
  ggtitle("Pairwise fraction of sites IBS") +
  theme_bw() + theme(legend.position = "none",
                     strip.text.x = element_text(size=10),
                     plot.title = element_text(hjust=0.5, face="bold"))

show(ibs_diagnostic_plot)

pdf("Empirical_results/Guyana_Colombia_ibs_diagnostic.pdf", height=2.5, width=10)
show(ibs_diagnostic_plot)
invisible(dev.off())

```

### Relatedness estimates based on allelic states vs IBS descriptives

We also compare estimates of relatedness under (n)IBD independence based on the standard (n)IBD-to-allele model vs standard nIBD-to-IBS model.

```{r allele_vs_ibs_plot, echo=FALSE, fig.height=3, fig.width=10}

allele_vs_IBS_diagnostic_plot <- ggplot(summary_metrics) + 
  geom_point(aes(x=rhat_indep_allele, y=rhat_indep_ibs, fill=comparison), 
             pch=21, alpha=0.1, size=1.1) + 
  geom_abline(col="#bf490a", lty=2, lwd=0.8) +
  facet_grid(cols=vars(comparison)) +
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) +
  scale_fill_manual(values=pop_comparison, name="Comparison") +
  xlab("Estimate under (n)IBD independence with the standard (n)IBD-to-allele model") + 
  ylab("Estimate under (n)IBD independence\nwith the standard nIBD-to-IBS model") + 
  ggtitle("Relatedness estimates under (n)IBD independence based on allelic states vs IBS descriptives") +
  theme_bw() + theme(legend.position = "none",
                     strip.text.x = element_text(size=10),
                     plot.title = element_text(hjust=0.5, face="bold"))

show(allele_vs_IBS_diagnostic_plot)

png("Empirical_results/Guyana_Colombia_r_allele_vs_ibs_diagnostic.png", height=3, width=10, units="in", res=300)
show(allele_vs_IBS_diagnostic_plot)
invisible(dev.off())

```

### Relatedness estimates based on allelic states vs IBS descriptives

We plot the dense data diagnostic in the presence of population structure.

```{r dense_data_diagnostic_pop_structure, echo=FALSE, fig.height=5, fig.width=10}
dense_data_diagnostic_plot <- summary_metrics %>% 
  rename(`Standard\n(n)IBD-to-allele\nmodel`=rhat_indep_allele, 
         `Standard\nnIBD-to-IBS\nmodel`=rhat_indep_ibs) %>% 
  select(-V1, -V2, -ibs, -khat_hmm_allele, -N_comparable) %>% 
  reshape2::melt(id=c("comparison", "rhat_hmm_allele")) %>% 
  ggplot() +
  geom_point(aes(x=rhat_hmm_allele, y=value, fill=comparison), 
             pch=21, alpha=0.1, size=1.1) + 
  geom_abline(col="#bf490a", lty=2, lwd=0.8) +
  facet_grid(cols=vars(comparison), rows=vars(variable)) +
  coord_cartesian(xlim=c(0, 1), ylim=c(0, 1)) +
  scale_fill_manual(values=pop_comparison, name="Comparison") +
  xlab("Estimate under HMM with standard (n)IBD-to-allele model") + 
  ylab("Estimate under (n)IBD independence") + 
  ggtitle("Dense data diagnostic") +
  theme_bw() + theme(legend.position = "none",
                     plot.title = element_text(hjust=0.5, face="bold"),
                     strip.text.x = element_text(size=10),
                     strip.text.y = element_text(face="bold", angle=0, size=10))

show(dense_data_diagnostic_plot)

png("Empirical_results/Guyana_Colombia_dense_data_diagnostic.png", height=5, width=10, units="in", res=300)
show(dense_data_diagnostic_plot)
invisible(dev.off())

```

### Zero inflation

In the presence of population structure, we check for zero inflation in estimates of relatedness under (n)IBD independence based on the standard (n)IBD-to-allele model vs standard nIBD-to-IBS model.

```{r zero_inflation_pop_structure, echo=FALSE, fig.height=7, fig.width=10}
zero_inflation_plot <- summary_metrics %>% 
  select(-V1, -V2, -ibs, -khat_hmm_allele, -N_comparable) %>% 
  rename(`(n)IBD independence,\nIBS descriptives`=rhat_indep_ibs,
         `(n)IBD independence,\nallelic states`=rhat_indep_allele,
         `HMM,\nallelic states`=rhat_hmm_allele) %>%
  reshape2::melt(id="comparison") %>% 
  ggplot(aes(x=value, fill=comparison)) + geom_histogram(bins=100, col="black", lwd=0.1) +
  scale_fill_manual(values=pop_comparison, name="Comparison") +
  facet_grid2(variable~comparison, scales="free_y", independent="y") +
  xlab("Relatedness estimate") + ylab("Frequency") +
  ggtitle("Pairwise relatedness estimates") +
  theme_bw() + theme(legend.position = "none",
                     strip.text.x = element_text(size=10),
                     strip.text.y = element_text(size=10, face="bold"),
                     plot.title = element_text(hjust=0.5, face="bold"))

show(zero_inflation_plot)

png("Empirical_results/Guyana_Colombia_zero_inflation.png", height=7, width=10, units="in", res=300)
show(zero_inflation_plot)
invisible(dev.off())
```

